# Семинар 2. Домашняя работа 2
## Задание 1

In [ ]:
import pandas as pd

### Считать данные с помощью pandas

In [8]:
df = pd.read_csv('kc_house_data.csv', index_col='id', parse_dates=['date'], dayfirst=False)
df.head()

date     price  bedrooms  bathrooms  sqft_living  sqft_lot  \
id                                                                            
7129300520 2014-10-13  221900.0         3       1.00         1180      5650   
6414100192 2014-12-09  538000.0         3       2.25         2570      7242   
5631500400 2015-02-25  180000.0         2       1.00          770     10000   
2487200875 2014-12-09  604000.0         4       3.00         1960      5000   
1954400510 2015-02-18  510000.0         3       2.00         1680      8080   

            floors  waterfront  view  condition  grade  sqft_above  \
id                                                                   
7129300520     1.0           0     0          3      7        1180   
6414100192     2.0           0     0          3      7        2170   
5631500400     1.0           0     0          3      6         770   
2487200875     1.0           0     0          5      7        1050   
1954400510     1.0           0     0          3      8        1680   

            sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
id                                                                             
7129300520              0      1955             0    98178  47.5112 -122.257   
6414100192            400      1951          1991    98125  47.7210 -122.319   
5631500400              0      1933             0    98028  47.7379 -122.233   
2487200875            910      1965             0    98136  47.5208 -122.393   
1954400510              0      1987             0    98074  47.6168 -122.045   

            sqft_living15  sqft_lot15  
id                                     
7129300520           1340        5650  
6414100192           1690        7639  
5631500400           2720        8062  
2487200875           1360        5000  
1954400510           1800        7503

In [13]:
df['date'].dt.month_name()

id
7129300520     October
6414100192    December
5631500400    February
2487200875    December
1954400510    February
                ...   
263000018          May
6600060120    February
1523300141        June
291310100      January
1523300157     October
Name: date, Length: 21613, dtype: object

In [16]:
df['date'].dt.year

id
7129300520    2014
6414100192    2014
5631500400    2015
2487200875    2014
1954400510    2015
              ... 
263000018     2014
6600060120    2015
1523300141    2014
291310100     2015
1523300157    2014
Name: date, Length: 21613, dtype: int64

#### Вывести на экран первые 5 строк

In [17]:
df.head()

date     price  bedrooms  bathrooms  sqft_living  sqft_lot  \
id                                                                            
7129300520 2014-10-13  221900.0         3       1.00         1180      5650   
6414100192 2014-12-09  538000.0         3       2.25         2570      7242   
5631500400 2015-02-25  180000.0         2       1.00          770     10000   
2487200875 2014-12-09  604000.0         4       3.00         1960      5000   
1954400510 2015-02-18  510000.0         3       2.00         1680      8080   

            floors  waterfront  view  condition  grade  sqft_above  \
id                                                                   
7129300520     1.0           0     0          3      7        1180   
6414100192     2.0           0     0          3      7        2170   
5631500400     1.0           0     0          3      6         770   
2487200875     1.0           0     0          5      7        1050   
1954400510     1.0           0     0          3      8        1680   

            sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
id                                                                             
7129300520              0      1955             0    98178  47.5112 -122.257   
6414100192            400      1951          1991    98125  47.7210 -122.319   
5631500400              0      1933             0    98028  47.7379 -122.233   
2487200875            910      1965             0    98136  47.5208 -122.393   
1954400510              0      1987             0    98074  47.6168 -122.045   

            sqft_living15  sqft_lot15  
id                                     
7129300520           1340        5650  
6414100192           1690        7639  
5631500400           2720        8062  
2487200875           1360        5000  
1954400510           1800        7503

In [10]:
df.shape

(21613, 20)

#### Изучите типы данных

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21613 entries, 7129300520 to 1523300157
Data columns (total 20 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   date           21613 non-null  datetime64[ns]
 1   price          21613 non-null  float64       
 2   bedrooms       21613 non-null  int64         
 3   bathrooms      21613 non-null  float64       
 4   sqft_living    21613 non-null  int64         
 5   sqft_lot       21613 non-null  int64         
 6   floors         21613 non-null  float64       
 7   waterfront     21613 non-null  int64         
 8   view           21613 non-null  int64         
 9   condition      21613 non-null  int64         
 10  grade          21613 non-null  int64         
 11  sqft_above     21613 non-null  int64         
 12  sqft_basement  21613 non-null  int64         
 13  yr_built       21613 non-null  int64         
 14  yr_renovated   21613 non-null  int64         
 15  zipco

#### Посмотреть на описание признаков и на их содержание

In [ ]:
df.dtypes

date              object
price            float64
bedrooms           int64
bathrooms        float64
sqft_living        int64
sqft_lot           int64
floors           float64
waterfront         int64
view               int64
condition          int64
grade              int64
sqft_above         int64
sqft_basement      int64
yr_built           int64
yr_renovated       int64
zipcode            int64
lat              float64
long             float64
sqft_living15      int64
sqft_lot15         int64
dtype: object

#### Найдите количество пропущенных ячеек в данных

In [ ]:
df.isnull().sum()

date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

#### Посчитайте основные статистики по всем признакам и поизучайте их

In [18]:
df.describe()

price      bedrooms     bathrooms   sqft_living      sqft_lot  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  2.161300e+04   
mean   5.400881e+05      3.370842      2.114757   2079.899736  1.510697e+04   
std    3.671272e+05      0.930062      0.770163    918.440897  4.142051e+04   
min    7.500000e+04      0.000000      0.000000    290.000000  5.200000e+02   
25%    3.219500e+05      3.000000      1.750000   1427.000000  5.040000e+03   
50%    4.500000e+05      3.000000      2.250000   1910.000000  7.618000e+03   
75%    6.450000e+05      4.000000      2.500000   2550.000000  1.068800e+04   
max    7.700000e+06     33.000000      8.000000  13540.000000  1.651359e+06   

             floors    waterfront          view     condition         grade  \
count  21613.000000  21613.000000  21613.000000  21613.000000  21613.000000   
mean       1.494309      0.007542      0.234303      3.409430      7.656873   
std        0.539989      0.086517      0.766318      0.650743      1.175459   
min        1.000000      0.000000      0.000000      1.000000      1.000000   
25%        1.000000      0.000000      0.000000      3.000000      7.000000   
50%        1.500000      0.000000      0.000000      3.000000      7.000000   
75%        2.000000      0.000000      0.000000      4.000000      8.000000   
max        3.500000      1.000000      4.000000      5.000000     13.000000   

         sqft_above  sqft_basement      yr_built  yr_renovated       zipcode  \
count  21613.000000   21613.000000  21613.000000  21613.000000  21613.000000   
mean    1788.390691     291.509045   1971.005136     84.402258  98077.939805   
std      828.090978     442.575043     29.373411    401.679240     53.505026   
min      290.000000       0.000000   1900.000000      0.000000  98001.000000   
25%     1190.000000       0.000000   1951.000000      0.000000  98033.000000   
50%     1560.000000       0.000000   1975.000000      0.000000  98065.000000   
75%     2210.000000     560.000000   1997.000000      0.000000  98118.000000   
max     9410.000000    4820.000000   2015.000000   2015.000000  98199.000000   

                lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000   21613.000000   21613.000000  
mean      47.560053   -122.213896    1986.552492   12768.455652  
std        0.138564      0.140828     685.391304   27304.179631  
min       47.155900   -122.519000     399.000000     651.000000  
25%       47.471000   -122.328000    1490.000000    5100.000000  
50%       47.571800   -122.230000    1840.000000    7620.000000  
75%       47.678000   -122.125000    2360.000000   10083.000000  
max       47.777600   -121.315000    6210.000000  871200.000000

#### **Выводы:**
#### В данных 21613 строк и 20 столбцов после удаления столбца индекса.
#### Все данные таблицы заполнены.
#### Столбцы и их значение:

- id - Уникальный ID для каждого дома
- date - Дата продажи дома
- price - Стоимость продажи дома
- bedrooms - Кол-во спален
- bathrooms - Кол-во ванных комнат (0.5 - туалет без душа)
- sqft_living - Кв. метры жилые
- sqft_lot - Кв. метры общие
- floors - Кол-во этажей
- waterfront - Есть набержная или нет
- view - Значение от 0 до 4 насколько хороший вид
- condition - Значение от 1 до 5 насколько хорошее состояние
- grade - Значение от 1 до 13, где 1-3 плохая конструкция здания и дизайн, 7 - средний уровень конструкции и дизайна, 11-13 - высокое качество конструкции и дизайна
- sqft_above - Кв. метры дома, которые находятся выше земли
- sqft_basement - Кв. метры дома, которые находятся ниже земли
- yr_built - Год постройки дома
- yr_renovated - Год ремонта дома
- zipcode - Индекс
- lat - Широтма
- long - Долгота
- sqft_living15 - Кв. метры жилой площади у 15 соседей
- sqft_lot15 - Кв. метры общей площади у 15 соседей



## Задание 2
### Ответьте на несколько вопросов
### 3.1 В каком диапазоне изменяются стоимости недвижимости?


In [ ]:
df['price'].min(), df['price'].max()

(75000.0, 7700000.0)

### 3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?

In [ ]:
df_sqft_rate = df['sqft_living'].mean() * 100 / df['sqft_lot'].mean()
df_sqft_rate

13.76781757959227

#### 3.3 Как много домов с разными этажами в данных?

In [ ]:
df_floors = df['floors'].value_counts()
df_floors

1.0    10680
2.0     8241
1.5     1910
3.0      613
2.5      161
3.5        8
Name: floors, dtype: int64

#### 3.4 Насколько хорошие состояния у домов в данных?

In [ ]:
df['condition'].value_counts()

3    14031
4     5679
5     1701
2      172
1       30
Name: condition, dtype: int64

#### 3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?

In [ ]:
df.sort_values('yr_built').iloc[0, [13]]

yr_built    1900
Name: 9126101645, dtype: object

In [ ]:
df.sort_values('yr_built', ascending=False).iloc[0, [13]]

yr_built    2015
Name: 923059259, dtype: object

In [19]:
df['yr_built'].min(), df['yr_built'].max()

(1900, 2015)

## Задание 3.
### Ответьте на несколько вопросов
### 4.1 Сколько в среднем стоят дома, у которых 2 спальни?


In [ ]:
df[df['bedrooms'] == 2]['price'].mean()

401372.681884058

#### 4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?

In [ ]:
(df[df['price'] > 600000]['sqft_lot'].mean()).round(2)

20442.52

#### 4.3 Как много домов коснулся ремонт?

In [ ]:
df_years = df[df['yr_renovated'] != 0]['yr_renovated'].value_counts()
df_years.sum()

914

In [ ]:
df[df['yr_renovated'] != 0].shape[0]

914

#### 4.4 Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?

In [ ]:
df_top = (df[df['grade'] > 10]['price'].mean()).round(2)
df_top

1678635.12

In [ ]:
df_low = (df[df['grade'] < 4]['price'].mean()).round(2)
df_low

189750.0

In [ ]:
df_top - df_low

1488885.12

In [21]:
(df[df['grade'] > 10]['price'].mean() - df[df['grade'] < 4]['price'].mean()).round(2)

1488885.12

## Задание 4.
### 5.1 Выберите дом клиенту. Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?

In [22]:
df[(df['waterfront'] != 0) & (df['bathrooms'] >= 3) & (df['sqft_basement'] != 0)].shape[0]

41

### 5.2 Выберите дом клиенту. Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?


In [ ]:
df_selection = ((df['view'] == 4) | (df['waterfront'] == 1)) & (df['condition'] == 5) & (df['yr_built'] >= 1980)
choice = df[df_selection]
choice['price'].min(), choice['price'].max()

(1295000.0, 3000000.0)

### 5.3 Выберите дом клиенту. Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

In [24]:
(df[(df['sqft_basement'] == 0) & (df['floors'] == 2) & (df['price'] < 150000)]['condition'].mean()).round(3)

2.833